File: studio/schemas/document.ts
export default {
  name: 'sogiceDocument',
  title: 'Document',
  type: 'document',
  fields: [

    // ── Workflow ────────────────────────────────────────────────
    {
      name: 'workflowStatus',
      title: 'Workflow Status',
      type: 'string',
      options: { list: ['unverified', 'in_progress', 'verified', 'published'] },
      initialValue: 'unverified',
    },
    {
      name: 'tier',
      title: 'Trust Tier',
      type: 'string',
      options: { list: ['1', '2', '3'] },
    },
    {
      name: 'tierAssignedBy',
      title: 'Tier Assigned By',
      type: 'string',
      options: { list: ['auto', 'researcher'] },
      initialValue: 'auto',
    },
    {
      name: 'tierChangeLog',
      title: 'Tier Change Log',
      type: 'array',
      of: [{
        type: 'object',
        fields: [
          { name: 'from', type: 'string' },
          { name: 'to', type: 'string' },
          { name: 'changedAt', type: 'datetime' },
          { name: 'reason', type: 'string' },
        ],
      }],
    },

    // ── Meta / Provenance ───────────────────────────────────────
    {
      name: 'meta',
      title: 'Meta',
      type: 'object',
      fields: [
        { name: 'sourceUrl', title: 'Source URL', type: 'url' },
        { name: 'archiveUrl', title: 'Wayback Archive URL', type: 'url' },
        { name: 'fileRef', title: 'File Asset', type: 'file' },
        { name: 'ingestedAt', title: 'Ingested At', type: 'datetime' },
        {
          name: 'ingestionBatch',
          title: 'Ingestion Batch',
          type: 'reference',
          to: [{ type: 'ingestionBatch' }],
        },
        {
          name: 'preprocessingTool',
          title: 'Preprocessing Tool',
          type: 'string',
          options: { list: ['unstructured', 'docling', 'yt-dlp', 'whisper', 'manual', 'none'] },
        },
        {
          name: 'preprocessingQuality',
          title: 'Preprocessing Quality',
          type: 'string',
          options: { list: ['high', 'medium', 'low', 'blocked'] },
        },
      ],
    },
    {
      name: 'provenance',
      title: 'Provenance',
      type: 'object',
      fields: [
        { name: 'originalUrl', type: 'url' },
        { name: 'accessedVia', type: 'string' },
        { name: 'waybackUrl', type: 'url' },
        { name: 'htmlSnapshotRef', type: 'file' },
        { name: 'chainNotes', type: 'text', rows: 2 },
      ],
    },

    // ── Classification ──────────────────────────────────────────
    {
      name: 'classification',
      title: 'Classification',
      type: 'object',
      fields: [
        { name: 'type', type: 'string' },
        { name: 'format', type: 'string' },
        { name: 'evidence', type: 'array', of: [{ type: 'string' }] },
        { name: 'scope', type: 'string', options: { list: ['Core', 'Contextual', 'Reference'] } },
        { name: 'country', type: 'array', of: [{ type: 'string' }] },
        { name: 'tactic', type: 'array', of: [{ type: 'string' }] },
        { name: 'actor', type: 'array', of: [{ type: 'string' }] },
        { name: 'network', type: 'array', of: [{ type: 'string' }] },
        { name: 'practice', type: 'array', of: [{ type: 'string' }] },
        { name: 'term', type: 'array', of: [{ type: 'string' }] },
        { name: 'harm', type: 'array', of: [{ type: 'string' }] },
        { name: 'migration', type: 'array', of: [{ type: 'string' }] },
        { name: 'function', type: 'array', of: [{ type: 'string' }] },
        { name: 'landmark', type: 'array', of: [{ type: 'string' }] },
        { name: 'flags', type: 'array', of: [{ type: 'string' }] },
        { name: 'narrativeRegister', type: 'string' },
      ],
    },

    // ── Confidence ──────────────────────────────────────────────
    {
      name: 'confidence',
      title: 'Confidence',
      type: 'object',
      fields: [
        { name: 'overallScore', title: 'Overall Score (0.0–1.0)', type: 'number' },
        {
          name: 'status',
          title: 'Status',
          type: 'string',
          options: { list: ['high', 'medium', 'low'] },
        },
        { name: 'reasons', title: 'Reasons', type: 'array', of: [{ type: 'string' }] },
        {
          name: 'signals',
          title: 'Signals',
          type: 'object',
          fields: [
            { name: 'textQuality', type: 'string', options: { list: ['clean', 'noisy'] } },
            { name: 'languageClarity', type: 'string', options: { list: ['clear', 'mixed', 'unclear'] } },
            { name: 'contentStructure', type: 'string', options: { list: ['well-structured', 'ambiguous'] } },
          ],
        },
      ],
    },

    // ── Field-Level Confidence ──────────────────────────────────
    {
      name: 'fieldConfidence',
      title: 'Field-Level Confidence',
      type: 'object',
      fields: [
        { name: 'type', type: 'number' },
        { name: 'format', type: 'number' },
        { name: 'tactic', type: 'number' },
        { name: 'term', type: 'number' },
        { name: 'actor', type: 'number' },
        { name: 'scope', type: 'number' },
        {
          name: 'lowConfidenceReasons',
          title: 'Low Confidence Reasons',
          type: 'array',
          of: [{
            type: 'object',
            fields: [
              { name: 'field', type: 'string' },
              { name: 'issue', type: 'string' },
              { name: 'severity', type: 'string', options: { list: ['low', 'medium', 'high'] } },
            ],
          }],
        },
      ],
    },

    // ── Document Date ───────────────────────────────────────────
    {
      name: 'documentDate',
      title: 'Document Date',
      type: 'object',
      fields: [
        { name: 'year', type: 'number' },
        { name: 'month', type: 'number' },
        { name: 'day', type: 'number' },
        {
          name: 'dateConfidence',
          title: 'Date Confidence',
          type: 'string',
          options: { list: ['exact', 'approximate', 'unknown'] },
        },
      ],
    },

    // ── Entity Links ────────────────────────────────────────────
    {
      name: 'entities',
      title: 'Entity Links',
      type: 'object',
      fields: [
        {
          name: 'actorsLinked',
          title: 'Actors (Persons)',
          type: 'array',
          of: [{ type: 'reference', to: [{ type: 'person' }] }],
        },
        {
          name: 'organizationsLinked',
          title: 'Organizations',
          type: 'array',
          of: [{ type: 'reference', to: [{ type: 'organization' }] }],
        },
        {
          name: 'networksLinked',
          title: 'Networks',
          type: 'array',
          of: [{ type: 'reference', to: [{ type: 'organization' }] }],
        },
        {
          name: 'lawsLinked',
          title: 'Laws / Policies',
          type: 'array',
          of: [{ type: 'reference', to: [{ type: 'lawPolicy' }] }],
        },
        {
          name: 'eventsLinked',
          title: 'Events',
          type: 'array',
          of: [{ type: 'reference', to: [{ type: 'event' }] }],
        },
        {
          name: 'legalDefinitionsLinked',
          title: 'Legal Definitions',
          type: 'array',
          of: [{ type: 'reference', to: [{ type: 'legalDefinition' }] }],
        },
        {
          name: 'exclusionClausesLinked',
          title: 'Exclusion Clauses',
          type: 'array',
          of: [{ type: 'reference', to: [{ type: 'exclusionClause' }] }],
        },
      ],
    },

    // ── Content ─────────────────────────────────────────────────
    {
      name: 'content',
      title: 'Content',
      type: 'object',
      fields: [
        { name: 'title', type: 'string' },
        { name: 'summary', title: 'Research Summary (80–150 words)', type: 'text', rows: 5 },
        { name: 'summaryValidation', title: 'Validation Summary', type: 'text', rows: 5 },
        { name: 'narrativeRegister', type: 'string' },
        { name: 'languageDetected', title: 'Language (ISO 639-1)', type: 'string' },
        { name: 'wordCount', type: 'number' },
        { name: 'extractedText', title: 'Extracted Text (not in public export)', type: 'text', rows: 20 },
      ],
    },

    // ── Referenced URLs ─────────────────────────────────────────
    {
      name: 'referencedUrls',
      title: 'Referenced URLs',
      type: 'array',
      of: [{
        type: 'object',
        fields: [
          { name: 'url', type: 'url' },
          { name: 'anchorText', type: 'string' },
          {
            name: 'linkType',
            type: 'string',
            options: { list: ['outbound', 'citation', 'source', 'related', 'unknown'] },
          },
          { name: 'domain', type: 'string' },
          { name: 'resolved', type: 'boolean', initialValue: false },
          { name: 'archiveUrl', type: 'url' },
        ],
      }],
    },

    // ── Social Media ────────────────────────────────────────────
    {
      name: 'socialMedia',
      title: 'Social Media',
      type: 'object',
      fields: [
        { name: 'accountName', type: 'string' },
        { name: 'followerCountAtCapture', type: 'number' },
        { name: 'hashtags', type: 'array', of: [{ type: 'string' }] },
        { name: 'partOfSeries', type: 'boolean', initialValue: false },
        { name: 'seriesId', type: 'string' },
      ],
    },

    // ── Priority Score ──────────────────────────────────────────
    {
      name: 'priorityScore',
      title: 'Priority Score (1–5 per axis)',
      type: 'object',
      fields: [
        { name: 'artistic', type: 'number' },
        { name: 'network', type: 'number' },
        { name: 'lexicon', type: 'number' },
        { name: 'testimony', type: 'number' },
        { name: 'historical', type: 'number' },
      ],
    },

    // ── Extractable Assets ──────────────────────────────────────
    {
      name: 'extractableAssets',
      title: 'Extractable Assets',
      type: 'array',
      of: [{
        type: 'object',
        fields: [
          {
            name: 'assetType',
            type: 'string',
            options: {
              list: [
                'prayer_script', 'testimony_excerpt', 'conversion_script',
                'course_structure', 'statistical_claim', 'network_connection',
                'terminology_coinage', 'visual_asset', 'legislative_quote', 'counter_sermon',
              ],
            },
          },
          { name: 'content', type: 'text', rows: 4 },
          { name: 'targetModule', type: 'string' },
          {
            name: 'extractedBy',
            type: 'string',
            options: { list: ['llm_primary', 'llm_validation', 'human'] },
          },
        ],
      }],
    },

    // ── Candidate Terms ─────────────────────────────────────────
    {
      name: 'candidateTerms',
      title: 'Candidate Terms',
      type: 'array',
      of: [{
        type: 'object',
        fields: [
          { name: 'term', type: 'string' },
          { name: 'language', title: 'Language (ISO 639-1)', type: 'string' },
          { name: 'proposedCategory', type: 'string' },
          { name: 'promotionalUse', type: 'boolean' },
          { name: 'draftDefinition', type: 'text', rows: 3 },
          { name: 'contextQuote', type: 'string' },
          { name: 'approved', type: 'boolean', initialValue: false },
          { name: 'newStatus', type: 'string', options: { list: ['draft', 'discarded'] } },
          {
            name: 'lexiconEntryRef',
            title: 'Lexicon Entry (if approved)',
            type: 'reference',
            to: [{ type: 'lexiconEntry' }],
          },
        ],
      }],
    },

    // ── Validation ──────────────────────────────────────────────
    {
      name: 'validation',
      title: 'Validation',
      type: 'object',
      fields: [
        {
          name: 'status',
          type: 'string',
          options: { list: ['not_validated', 'queued', 'validated'] },
          initialValue: 'not_validated',
        },
        {
          name: 'validationTrigger',
          type: 'string',
          options: {
            list: [
              'mandatory_threshold', 'low_confidence', 'legal_flag',
              'testimony_flag', 'tier_requirement', 'manual_researcher', 'random_audit',
            ],
          },
        },
        { name: 'preValidationConfidence', type: 'number' },
        { name: 'postValidationConfidence', type: 'number' },
        { name: 'validatedBy', title: 'Validated By (model identifier)', type: 'string' },
        { name: 'validatedAt', type: 'datetime' },
        {
          name: 'resolution',
          type: 'string',
          options: { list: ['accepted_primary', 'accepted_validation', 'human_override'] },
        },
        {
          name: 'validationBatchRef',
          title: 'Validation Batch',
          type: 'reference',
          to: [{ type: 'validationBatch' }],
        },
      ],
    },

    // ── AI Metadata ─────────────────────────────────────────────
    {
      name: 'aiMetadata',
      title: 'AI Metadata',
      type: 'object',
      fields: [
        { name: 'primaryModel', type: 'string' },
        {
          name: 'primaryProvider',
          type: 'string',
          options: { list: ['anthropic', 'openai', 'google', 'local', 'other'] },
        },
        { name: 'validationModel', type: 'string' },
        { name: 'validationProvider', type: 'string' },
        {
          name: 'promptVersionRef',
          title: 'Prompt Version',
          type: 'reference',
          to: [{ type: 'promptVersion' }],
        },
        { name: 'ontologyVersion', type: 'string', initialValue: 'v3.0' },
        { name: 'processingDate', type: 'datetime' },
        { name: 'inputLengthChars', type: 'number' },
        { name: 'truncated', type: 'boolean', initialValue: false },
        {
          name: 'agreementStatus',
          type: 'string',
          options: { list: ['agreed', 'disagreed', 'not_validated'] },
          initialValue: 'not_validated',
        },
        { name: 'disagreements', type: 'array', of: [{ type: 'string' }] },
        {
          name: 'resolution',
          type: 'string',
          options: { list: ['accepted_primary', 'accepted_validation', 'human_override', 'not_applicable'] },
          initialValue: 'not_applicable',
        },
        {
          name: 'humanReview',
          title: 'Human Review',
          type: 'object',
          fields: [
            { name: 'reviewedBy', type: 'string', options: { list: ['researcher', 'intern'] } },
            { name: 'reviewedAt', type: 'datetime' },
            { name: 'changesMade', type: 'boolean' },
          ],
        },
        {
          name: 'algorithmicPreprocessing',
          title: 'Algorithmic Preprocessing',
          type: 'object',
          fields: [
            {
              name: 'tool',
              type: 'string',
              options: { list: ['unstructured', 'docling', 'yt-dlp', 'whisper', 'manual', 'none'] },
            },
            {
              name: 'preprocessingQuality',
              type: 'string',
              options: { list: ['high', 'medium', 'low', 'blocked'] },
            },
          ],
        },
      ],
    },

    // ── Zotero ──────────────────────────────────────────────────
    {
      name: 'zotero',
      title: 'Zotero',
      type: 'object',
      fields: [
        { name: 'collection', type: 'string' },
        { name: 'exportedAt', type: 'datetime' },
      ],
    },

    { name: 'testimonyFlag', title: 'Testimony Flag', type: 'boolean', initialValue: false },
    { name: 'needsReview', title: 'Needs Review', type: 'boolean', initialValue: false },
  ],

  preview: {
    select: {
      title: 'content.title',
      subtitle: 'workflowStatus',
      media: 'meta.fileRef',
    },
  },
}

File: runner/models/document.py
"""
Pydantic models that mirror the ingestion-v3.1 output schema exactly.
AnalysisResult validates Claude's or Ollama's JSON response.
IntakeResult and PreprocessResult carry pipeline state between stages.
"""
from __future__ import annotations
from dataclasses import dataclass, field
from pathlib import Path
from typing import Literal, Optional
from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# ingestion-v3.1 output schema
# ---------------------------------------------------------------------------

class DocumentDate(BaseModel):
    year: int = 0
    month: int = 0
    day: int = 0
    confidence: Literal["exact", "approximate", "unknown"] = "unknown"


class Priority(BaseModel):
    artistic: int = Field(default=1, ge=1, le=5)
    network: int = Field(default=1, ge=1, le=5)
    lexicon: int = Field(default=1, ge=1, le=5)
    testimony: int = Field(default=1, ge=1, le=5)
    historical: int = Field(default=1, ge=1, le=5)


class ConfidenceSignals(BaseModel):
    text_quality: Literal["clean", "noisy"] = "clean"
    language_clarity: Literal["clear", "mixed", "unclear"] = "clear"
    content_structure: Literal["well-structured", "ambiguous"] = "well-structured"


class Confidence(BaseModel):
    overall_score: float = Field(default=0.0, ge=0.0, le=1.0)
    status: Literal["high", "medium", "low"] = "low"
    reasons: list[str] = Field(default_factory=list)
    signals: ConfidenceSignals = Field(default_factory=ConfidenceSignals)


class LowConfidenceReason(BaseModel):
    field: str
    issue: str
    severity: Literal["low", "medium", "high"] = "low"


class FieldConfidence(BaseModel):
    type: float = Field(default=1.0, ge=0.0, le=1.0)
    format: float = Field(default=1.0, ge=0.0, le=1.0)
    tactic: float = Field(default=1.0, ge=0.0, le=1.0)
    term: float = Field(default=1.0, ge=0.0, le=1.0)
    actor: float = Field(default=1.0, ge=0.0, le=1.0)
    scope: float = Field(default=1.0, ge=0.0, le=1.0)
    low_confidence_reasons: list[LowConfidenceReason] = Field(default_factory=list)


class CandidateTerm(BaseModel):
    term: str
    language: str = "unknown"
    proposed_category: str = ""
    promotional_use: bool = True
    draft_definition: str = ""
    context_quote: str = ""


class SuggestedActor(BaseModel):
    name: str
    type: Literal["person", "organization"] = "organization"
    country: str = ""
    role: str = ""
    evidence_quote: str = ""


class SuggestedNetwork(BaseModel):
    name: str
    description: str = ""
    evidence_quote: str = ""


class ExtractableAsset(BaseModel):
    asset_type: Literal[
        "prayer_script", "testimony_excerpt", "conversion_script",
        "course_structure", "statistical_claim", "network_connection",
        "terminology_coinage", "visual_asset", "legislative_quote", "counter_sermon",
    ] = "statistical_claim"
    content: str = ""
    target_module: str = ""
    extracted_by: str = "llm_primary"


DocumentType = Literal[
    "Pro-SOGICE", "Anti-SOGICE", "Neutral-Academic", "Legal-Instrument",
    "Testimony", "Media-Coverage", "Internal-Org-Document", "Mixed",
]

DocumentFormat = Literal[
    "Website-Page", "Blog-Post", "Social-Media-Post", "Video", "Podcast",
    "Academic-Paper", "NGO-Report", "Government-Report", "Court-Judgment",
    "Legislative-Submission", "Parliamentary-Debate", "Press-Release",
    "Book", "Book-Chapter", "Pamphlet", "Newsletter", "Email",
    "Manual", "Course-Material", "Event-Program", "Other",
]

NarrativeRegister = Literal[
    "Pastoral-Healing", "Scientific-Clinical", "Legal-Policy", "Testimonial-Personal",
    "Conspiratorial", "Activist-Advocacy", "Journalistic", "Academic-Analytical", "Mixed",
]

Scope = Literal["Core", "Contextual", "Reference"]


class AnalysisResult(BaseModel):
    """Exact mirror of the ingestion-v3.1 JSON output schema."""
    model_config = ConfigDict(extra="ignore")

    type: DocumentType
    format: DocumentFormat
    evidence: list[str]
    scope: Scope
    country: list[str] = Field(default_factory=list)
    tactic: list[str] = Field(default_factory=list)
    actor: list[str] = Field(default_factory=list)
    network: list[str] = Field(default_factory=list)
    practice: list[str] = Field(default_factory=list)
    term: list[str] = Field(default_factory=list)
    harm: list[str] = Field(default_factory=list)
    migration: list[str] = Field(default_factory=list)
    function: list[str] = Field(default_factory=list)
    landmark: list[str] = Field(default_factory=list)
    flags: list[str] = Field(default_factory=list)
    narrative_register: NarrativeRegister
    document_date: DocumentDate = Field(default_factory=DocumentDate)
    summary: str
    priority: Priority = Field(default_factory=Priority)
    testimony_flag: bool = False
    needs_review: bool = False
    confidence: Confidence = Field(default_factory=Confidence)
    field_confidence: FieldConfidence = Field(default_factory=FieldConfidence)
    candidate_terms: list[CandidateTerm] = Field(default_factory=list)
    suggested_actors: list[SuggestedActor] = Field(default_factory=list)
    suggested_networks: list[SuggestedNetwork] = Field(default_factory=list)
    extractable_assets: list[ExtractableAsset] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Pipeline state models (passed between pipeline stages)
# ---------------------------------------------------------------------------

@dataclass
class PageIntelligence:
    """Rich web-page intelligence extracted at preprocessing time.
    All fields are optional — populated only when the source is a URL."""

    # Canonical / deduplication
    canonical_url: str = ""

    # Open Graph / social card metadata (often richer than page content)
    og_title: str = ""
    og_description: str = ""
    og_image: str = ""
    og_type: str = ""                       # article | website | video | ...

    # CMS taxonomy
    tags: list[str] = field(default_factory=list)
    categories: list[str] = field(default_factory=list)
    keywords: list[str] = field(default_factory=list)

    # Author enrichment
    author_url: str = ""                    # link to author profile page

    # Social media profiles found on the page (header/footer/share buttons)
    social_profiles: list[dict] = field(default_factory=list)
    # [{platform: "twitter", url: "https://twitter.com/CConcern", handle: "CConcern"}]

    # Document/file links — future ingestion queue
    document_links: list[dict] = field(default_factory=list)
    # [{url, file_type: "pdf"|"docx"|"pptx"..., anchor_text, domain}]

    # Embedded media — future ingestion queue
    media_embeds: list[dict] = field(default_factory=list)
    # [{platform: "youtube"|"vimeo"|"rumble", id, url, title?}]

    # Contact / org identity signals
    emails: list[str] = field(default_factory=list)

    # JSON-LD structured data (Schema.org) — raw parsed objects
    json_ld: list[dict] = field(default_factory=list)

    # Internal links (same domain) — topical signature of the organisation
    internal_links: list[dict] = field(default_factory=list)
    # [{url, anchor_text}] — top 30

    # Outbound links (external domains)
    outbound_links: list[dict] = field(default_factory=list)
    # [{url, anchor_text, domain}] — top 80


@dataclass
class IntakeResult:
    doc_id: str
    source: str                            # original URL or file path string
    source_type: Literal["url", "pdf", "html", "video", "audio", "epub", "srt"]
    declared_type: str                     # user-supplied or auto-detected document type
    tier: int                              # 1 | 2 | 3
    batch_id: str
    language: Optional[str]               # ISO 639-1, None if unknown at intake
    archive_url: Optional[str] = None     # Wayback Machine URL once archived
    local_dir: Optional[Path] = None      # ~/survivingsogice/corpus/{doc_id}/


@dataclass
class PreprocessResult:
    doc_id: str
    tool_used: str                         # "docling" | "unstructured" | "trafilatura" | "whisper" | "manual"
    quality: Literal["high", "medium", "low", "blocked"]
    text: str                              # clean extracted text (primary context for LLM)
    markdown: Optional[str] = None        # structure-preserving markdown (Docling output)
    ocr_images: list[dict] = field(default_factory=list)
    char_count: int = 0
    truncated: bool = False
    language_detected: Optional[str] = None
    # Rich provenance metadata (populated by trafilatura / docling)
    title: str = ""
    author: str = ""
    date_published: str = ""
    sitename: str = ""                     # publisher / organisation name as found on the page
    description: str = ""                  # meta description or lede
    hostname: str = ""                     # bare domain, e.g. christianconcern.com
    outbound_links: list[dict] = field(default_factory=list)   # [{url, anchor_text, domain}]
    page_intel: Optional["PageIntelligence"] = None


@dataclass
class DocumentPackage:
    """Everything assembled after all pipeline stages complete."""
    intake: IntakeResult
    preprocess: PreprocessResult
    analysis: AnalysisResult
    embedding: list[float]
    embedding_model: str
    llm_used: str                          # "claude" | "gemma4:e4b" | "both"
    local_dir: Path
